### PR TITLE
feat: implementation of command send process

### PR DESCRIPTION
**Pull request by v-sugahara7993-esol**
## Description

This PR adds a process for nodes with services that require sending commands to facilities.

See [NODE-GRAPH(COMPANY INTERNAL LINK)](https://tier4.atlassian.net/wiki/spaces/AIP/pages/2674065672/) for messages between nodes. See [Cargo Loading Service Module(COMPANY INTERNAL LINK)](https://tier4.atlassian.net/wiki/spaces/AIP/pages/2681569626/EVPJ-107) for the design of modules containing nodes.

Also, the following remarks remained in the previous pull request have been corrected.
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107921356
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107922027
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107924613
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107925149
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107925850
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107938391
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107943175
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107944728
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107945292
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107951148
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107952195
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107953611
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107954015
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107954815
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107955597
* https://github.com/eve-autonomy/cargo_loading_service/pull/1#discussion_r1107959533

TODO:
* The part where state and status are different in naming will be discussed and corrected after the 2nd make's PR has run its course.
* Optional tag value check process not yet implemented.

## Related links

* Message Design
   * [NODE-GRAPH(COMPANY INTERNAL LINK)](https://tier4.atlassian.net/wiki/spaces/AIP/pages/2674065672/)
* Module Design
   * [Cargo Loading Service Module(COMPANY INTERNAL LINK)](https://tier4.atlassian.net/wiki/spaces/AIP/pages/2681569626/EVPJ-107)
* Jira ticket
   * [Ticket(COMPANY INTERNAL LINK)](https://tier4.atlassian.net/browse/T4PB-25941)

## Tests performed

The tests described below were conducted.

* Verified that the package can be built without Error/Warning by `colcon`.
* We started the node and waited for more than 1 minute to confirm that it did not CoreDump.
* Verified the input/output values of the nodes in the description of [Cargo Loading Service Module(COMPANY INTERNAL LINK)](https://tier4.atlassian.net/wiki/spaces/AIP/pages/2681569626/EVPJ-107).
   * Confirmation details are listed in Cargo Loading Service Checkpoint at the bottom.
   * The input/output values to be checked are listed in Cargo Loading Service Test Target at the bottom.
   * Details of the test can be found in ". /test/cargo_loading_service_test.md" for details.
* Verified that the node can communicate for the description of [NODE-GRAPH(COMPANY INTERNAL LINK)](https://tier4.atlassian.net/wiki/spaces/AIP/pages/2674065672/).
   * The details of the checks, the target to be checked, and the test are the same as for the check on the [Cargo Loading Service Module(COMPANY INTERNAL LINK)](https://tier4.atlassian.net/wiki/spaces/AIP/pages/2681569626/EVPJ-107) one above.
      * See Cargo Loading Service Checkpoint and Cargo Loading Service Test Target and 
       ". /test/cargo_loading_service_test.md".
* We have confirmed the startup of the node using the launch file.

<details><summary>Cargo Loading Service Checkpoint</summary>
<p>

|Check Point No.|Description|
|---|---|
|1|The service `/parking/cargo_loading` returns success or failure as the response value.|
|2|Service behavior changes depending on the value of `aw_state` received at `/in_parking/state`.|
|3|Service behavior changes depending on the value of `approval` received in `/infrastructure_status`.|
|4|Service behavior changes depending on the value of `id` received in `/infrastructure_status`.|

</p>
</details>

<details><summary>Node Communication Test Target</summary>
<p>

|Test Target No.|Test Target|Check Point No.|
|---|---|---|
|1 | Return SUCCESS(=1)                               |1|
|2 | Return FAIL(=2)                                  |1|
|3 | aw_state is NONE(=0)                             |2|
|4 | aw_state is AW_EMERGENCY(=1)                     |2|
|5 | aw_state is AW_OUT_OF_PARKING(=2)                |2|
|6 | aw_state is AW_ARRIVED_PARKING(=3)               |2|
|7 | aw_state is AW_INVALID_DEPARTURE_INSTRUCTION(=4) |2|
|8 | aw_state is AW_WAIT_DEPARTURE_INSTRUCTION(=5)    |2|
|9 | aw_state is AW_UNAVAILABLE(=6)                   |2|
|10| approval is true                                 |3|
|11| approval is false                                |3|
|12| Match with ID requested by service               |4|
|13| Mismatch with ID requested by service            |4|

</p>
</details>

## Notes for reviewers

The following outstanding items are included. Additional PR will be added before this feature is released.

* The function that a node should have.
* Description of README.md.
* Correction of remarks by Formatter and Linter.

## Pre-review checklist for the PR author

The PR author **must** check the checkboxes below when creating the PR.

- [x] I've confirmed the [contribution guidelines].
- [x] The PR follows the [pull request guidelines].

## In-review checklist for the PR reviewers

The PR reviewers **must** check the checkboxes below before approval.

- [x] The PR follows the [pull request guidelines].
- [x] The PR has been properly tested.
- [x] The PR has been reviewed by the code owners.

## Post-review checklist for the PR author

The PR author **must** check the checkboxes below before merging.

- [x] There are no open discussions or they are tracked via tickets.
- [x] The PR is ready for merge.

After all checkboxes are checked, anyone who has write access can merge the PR.

[contribution guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/
[pull request guidelines]: https://autowarefoundation.github.io/autoware-documentation/main/contributing/pull-request-guidelines/

### CODE REVIEW

File: include/cargo_loading_service/cargo_loading_service.hpp [hunk 1]
```diff
@@ -39,18 +39,17 @@ class CargoLoadingService : public rclcpp::Node
...
-
-  std::mutex mutex_cargo_loading_state_;
-  std::mutex mutex_parking_state_;
-  struct FacilityInfo {
-    std::string id;
-    bool approval;
-  } facility_info_;
+  // constants
+  static constexpr uint8_t CMD_STATE_REQUESTING = 0b01;
+  static constexpr uint8_t CMD_STATE_ERROR = 0b10;
```
Comment (h-ohta) [inline include/cargo_loading_service/cargo_loading_service.hpp:44]: enum classで実装したほうが良さそう？

File: src/cargo_loading_service.cpp [hunk 2]
```diff
@@ -22,17 +22,15 @@ namespace cargo_loading_service
 CargoLoadingService::CargoLoadingService(
   const rclcpp::NodeOptions & options = rclcpp::NodeOptions())
 : Node("cargo_loading_service", options),
-  facility_info_({"", false}),
-  aw_state_(in_parking_msgs::msg::InParkingStatus::AW_NONE)
+  facility_id_(""),
+  aw_state_(in_parking_msgs::msg::InParkingStatus::NONE)
```
Comment (h-ohta) [inline src/cargo_loading_service.cpp:26]: クラス側で{}初期化

File: src/cargo_loading_service.cpp [hunk 3]
```diff
@@ -74,87 +72,105 @@ void CargoLoadingService::execCargoLoading(
...
+  const int32_t finalizing_pub_limit = static_cast<int32_t>(command_pub_hz_ * 2.0);
+  int32_t finalizing_pub_count = 0;
+
+  facility_id_ = request->value;
+
+  response->state = ExecuteInParkingTaskResponse::SUCCESS;
+
+  const auto period = std::chrono::duration_cast<std::chrono::nanoseconds>(
+    std::chrono::duration<double>(1.0 / command_pub_hz_));
+  timer_ = create_timer(
```
Comment (h-ohta) [inline src/cargo_loading_service.cpp]: rclcppのtimerを使っているのはなにか理由があるんでしょうか？

Comment (h-ohta) [reply src/cargo_loading_service.cpp]: callback groupによりaw_state_へのアクセスの競合を防ぐため？

File: include/cargo_loading_service/cargo_loading_service.hpp [hunk 4]
```diff
@@ -63,15 +62,20 @@ class CargoLoadingService : public rclcpp::Node
...
-  rclcpp::Subscription<InParkingStatus>::SharedPtr sub_parking_state_;
+  rclcpp::Subscription<InParkingStatus>::SharedPtr sub_inparking_status_;
   rclcpp::Subscription<InfrastructureStateArray>::SharedPtr sub_cargo_loading_state_;
 
   // Callback
   void execCargoLoading(
     const ExecuteInParkingTask::Request::SharedPtr request,
     const ExecuteInParkingTask::Response::SharedPtr response);
-  void onParkingState(const InParkingStatus::ConstSharedPtr msg_ptr);
+  void onTimer();
```
Comment (h-ohta) [inline include/cargo_loading_service/cargo_loading_service.hpp:72]: 不要？

File: src/cargo_loading_service.cpp (hunk 3 above)
Comment (h-ohta) [reply src/cargo_loading_service.cpp]: https://github.com/eve-autonomy/cargo_loading_service/pull/1/files#r1107938391
この指摘と認識

File: include/cargo_loading_service/cargo_loading_service.hpp [hunk 5]
```diff
@@ -39,18 +39,17 @@ class CargoLoadingService : public rclcpp::Node
...
-    std::string id;
-    bool approval;
-  } facility_info_;
+  // constants
+  static constexpr uint8_t CMD_STATE_REQUESTING = 0b01;
+  static constexpr uint8_t CMD_STATE_ERROR = 0b10;
+  const std::string CMD_TYPE = "eva_beacon_system";
+
+  // variable
+  std::string facility_id_;
```
Comment (h-ohta) [inline include/cargo_loading_service/cargo_loading_service.hpp]: `#include <string>` をつける

File: include/cargo_loading_service/cargo_loading_service.hpp [hunk 6]
```diff
@@ -39,18 +39,17 @@ class CargoLoadingService : public rclcpp::Node
...
+  rclcpp::TimerBase::SharedPtr timer_;
```
Comment (h-ohta) [inline include/cargo_loading_service/cargo_loading_service.hpp:49]: timer変数がサービス内で完結するならメンバ変数は必要ない

File: src/cargo_loading_service.cpp [hunk 7]
```diff
@@ -22,17 +22,15 @@ namespace cargo_loading_service
 CargoLoadingService::CargoLoadingService(
   const rclcpp::NodeOptions & options = rclcpp::NodeOptions())
```
Comment (h-ohta) [inline src/cargo_loading_service.cpp]:
```suggestion
  const rclcpp::NodeOptions & options)
```

File: include/cargo_loading_service/cargo_loading_service.hpp [hunk 8]
```diff
@@ -39,18 +39,17 @@ class CargoLoadingService : public rclcpp::Node
...
   int32_t aw_state_;
```
Comment (h-ohta) [inline include/cargo_loading_service/cargo_loading_service.hpp:50]:
```suggestion
  int32_t aw_state_{InParkingStatus::NONE};
```

File: src/cargo_loading_service.cpp (hunk 2 above)
Comment (h-ohta) [inline src/cargo_loading_service.cpp:26]:
```suggestion
```

File: src/cargo_loading_service.cpp [hunk 9]
```diff
@@ -22,17 +22,15 @@ namespace cargo_loading_service
...
 : Node("cargo_loading_service", options),
-  facility_info_({"", false}),
-  aw_state_(in_parking_msgs::msg::InParkingStatus::AW_NONE)
+  facility_id_(""),
```
Comment (h-ohta) [inline src/cargo_loading_service.cpp:25]:
```suggestion
```

File: include/cargo_loading_service/cargo_loading_service.hpp (hunk 5 above)
Comment (h-ohta) [inline include/cargo_loading_service/cargo_loading_service.hpp:48]:
```suggestion
  std::string facility_id_{""};
```

File: src/cargo_loading_service.cpp [hunk 10]
```diff
@@ -74,87 +72,105 @@ void CargoLoadingService::execCargoLoading(
   using InParkingStatus = in_parking_msgs::msg::InParkingStatus;
   using ExecuteInParkingTaskResponse = in_parking_msgs::srv::ExecuteInParkingTask::Response;
 
-  uint8_t response_state = ExecuteInParkingTaskResponse::SUCCESS;
-  bool cmd_pub = true;
-  int32_t aw_state;
+  const int32_t finalizing_pub_limit = static_cast<int32_t>(command_pub_hz_ * 2.0);
```
Comment (h-ohta) [inline src/cargo_loading_service.cpp]:
```suggestion
  const auto finalizing_pub_limit = static_cast<int32_t>(command_pub_hz_ * 2.0);
```

File: include/cargo_loading_service/cargo_loading_service.hpp (hunk 5 above)
Comment (h-ohta) [reply include/cargo_loading_service/cargo_loading_service.hpp]: https://github.com/eve-autonomy/cargo_loading_service/pull/2/commits/ef650dc5f7b93a5b13a97bc823463f3723ac3a1f

File: include/cargo_loading_service/cargo_loading_service.hpp [hunk 11]
```diff
@@ -63,15 +64,20 @@ class CargoLoadingService : public rclcpp::Node
...
-  rclcpp::Subscription<InParkingStatus>::SharedPtr sub_parking_state_;
+  rclcpp::Subscription<InParkingStatus>::SharedPtr sub_inparking_status_;
   rclcpp::Subscription<InfrastructureStateArray>::SharedPtr sub_cargo_loading_state_;
 
   // Callback
   void execCargoLoading(
     const ExecuteInParkingTask::Request::SharedPtr request,
     const ExecuteInParkingTask::Response::SharedPtr response);
-  void onParkingState(const InParkingStatus::ConstSharedPtr msg_ptr);
+  void onTimer();
```
Comment (h-ohta) [inline include/cargo_loading_service/cargo_loading_service.hpp:77]:
```suggestion
```

File: src/cargo_loading_service.cpp (hunk 3 above)
Comment (yn-mrse) [reply src/cargo_loading_service.cpp]: serviceのみを別callback groupに設定することで、できるだけ排他不備による不具合が起きにくい構造にする狙いです。

**Comment by h-ohta** [comment]
| No  | テストケース                                                                                                 | 想定挙動                                                                             | 
| --- | ------------------------------------------------------------------------------------------------------------ | ------------------------------------------------------------------------------------ | 
| 1   | サービスが来ない                                                                                             | command: publishしない                                                               | 
| 2   | サービスが来た<br>- facility_idが0                                                                           | command: publishしない                                                               | 
| 3   | サービスが来た<br>- aw_stateがNONE                                                                           | command: publishしない<br>サービス: FAILを返す                                       | 
| 4   | サービスが来た<br>-  設備連携完了でない<br>- aw_stateがAW_EMERGENCY                                          | command: publish<br>サービス: 何も返さず継続                                         | 
| 5   | サービスが来た<br>-  設備連携完了でない<br>- aw_stateがAW_OUT_OF_PARKING<br>- aw_stateがAW_UNAVAILABLE<br>   | command: SEND_ZEROをpublish<br>サービス: FAILを返す                                  | 
| 6   | サービスが来た<br>-  設備連携完了でない<br>- aw_stateがAW_ARRIVED_PARKING<br>- aw_stateがAW_STILL_IN_PARKING | command: REQUESTINGをpublish<br>サービス: SUCCCESSを返す<br>サービス: 何も返さず継続 | 
| 7   | サービスが来た<br>-  設備連携完了                                                                            | command: SEND_ZEROをn秒間publish<br>サービス: SUCCESSを返す                          |

File: src/cargo_loading_service.cpp [hunk 12]
```diff
@@ -12,150 +12,178 @@
...
-  using ExecuteInParkingTaskResponse = in_parking_msgs::srv::ExecuteInParkingTask::Response;
+  // 設備ID取得
+  facility_id_ = request->value;
+
+  // aw_stateのsubscribe確認、NONEならばreject
+  if (aw_state_ == InParkingStatus::NONE) {
+    RCLCPP_WARN(this->get_logger(), "aw_state is NONE, reject...");
+    response->state = ExecuteInParkingTask::Response::FAIL;
+    return;
+  }
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
```
排他を考えるとonTimer内に寄せたほうが良いと思います。

File: src/cargo_loading_service.cpp [hunk 13]
```diff
@@ -12,150 +12,178 @@
...
 
-  // Create command
-  InfrastructureCommand command;
-  command.type = "eva_beacon_system";
-  command.id = request->value;
+  // キャンセルになるまで設備連携要求を投げ続ける
+  while (!timer_->is_canceled()) {
+    RCLCPP_INFO_THROTTLE(
+      this->get_logger(), *this->get_clock(), 1000 /* ms */, "request is running");
+  }
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
  // キャンセルになるまで設備連携要求を投げ続ける
  while (1) {
    rclcpp::sleep_for(rclcpp::Rate(command_pub_hz_).period());
    RCLCPP_INFO_THROTTLE(
      this->get_logger(), *this->get_clock(), 1000 /* ms */, "request is running");
    std::lock_guard<std::mutex> lock(mutex_);
    if (timer_->is_canceled()) {
      break;
    }
  }
```
- sleepしないとCPU負荷がやばいので追加をお願いします。
- timer_の排他制御に対策が必要と思います。

**Review by yn-mrse — CHANGES_REQUESTED** [review]

File: src/cargo_loading_service.cpp [hunk 14]
```diff
@@ -12,150 +12,178 @@
...
+    RCLCPP_INFO(this->get_logger(), "finished");
+    // SEND_ZEROをn秒間発出し、設備連携結果はSUCCESSで返し、timerをキャンセル
+    const auto start_time = this->now();
+    while (true) {
+      publishCommand(InfrastructureCommand::SEND_ZERO);
+      const auto time_diff = this->now() - start_time;
+      if (time_diff.seconds() > post_processing_time_) break;
+      rclcpp::sleep_for(rclcpp::Rate(command_pub_hz_).period());
+    }
+    service_result_ = ExecuteInParkingTask::Response::SUCCESS;
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
    infra_approval_ = false;
```

File: src/cargo_loading_service.cpp [hunk 15]
```diff
@@ -12,150 +12,178 @@
...
-  const in_parking_msgs::srv::ExecuteInParkingTask::Response::SharedPtr response)
+  const ExecuteInParkingTask::Request::SharedPtr request,
+  const ExecuteInParkingTask::Response::SharedPtr response)
 {
-  using InfrastructureCommand = v2i_interface_msgs::msg::InfrastructureCommand;
-  using InfrastructureCommandArray = v2i_interface_msgs::msg::InfrastructureCommandArray;
-  using InParkingStatus = in_parking_msgs::msg::InParkingStatus;
-  using ExecuteInParkingTaskResponse = in_parking_msgs::srv::ExecuteInParkingTask::Response;
+  // 設備ID取得
+  facility_id_ = request->value;
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
  // 設備ID取得
  infra_id_ = request->value;
  service_result_ = ExecuteInParkingTask::Response::SUCCESS;
```

File: src/cargo_loading_service.cpp [hunk 16]
```diff
@@ -12,150 +12,178 @@
...
+  response->state = service_result_;
 
-  {
-    std::lock_guard<std::mutex> lock(mutex_cargo_loading_state_);
-    facility_info_.approval = false;
-  }
+  // 各変数の初期化
+  service_result_ = ExecuteInParkingTask::Response::NONE;
+  facility_id_ = "";
+  infra_approval_ = false;
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp:101]:
```suggestion
```

File: src/cargo_loading_service.cpp [hunk 17]
```diff
@@ -12,150 +12,178 @@
...
+        publishCommand(static_cast<std::underlying_type<CommandState>::type>(CommandState::ERROR));
+        RCLCPP_ERROR(this->get_logger(), "AW emergency");
+        break;
+      // AWが停留所外などではSEND_ZEROを発出し、設備連携結果はFAILで返す
+      case InParkingStatus::AW_OUT_OF_PARKING:
+      case InParkingStatus::AW_UNAVAILABLE:
+        RCLCPP_WARN(this->get_logger(), "AW warning");
+        publishCommand(InfrastructureCommand::SEND_ZERO);
+        service_result_ = ExecuteInParkingTask::Response::FAIL;
+        timer_->cancel();
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
        infra_approval_ = true;
        service_result_ = ExecuteInParkingTask::Response::FAIL;
```

**Review by yn-mrse — CHANGES_REQUESTED** [review]
OUT_OF_PARKING等で一瞬だけ0b00を送る仕様を修正したいです。
下記のsuggested change反映で治るかと思います。
（ 関連：https://github.com/eve-autonomy/cargo_loading_service/pull/2#discussion_r1114236118 ）

File: src/cargo_loading_service.cpp [hunk 18]
```diff
@@ -12,150 +12,178 @@
...
+    // aw_stateで条件分岐
+    switch (aw_state_) {
+      // AWがEmergencyの場合はERRORを発出し続ける
+      case InParkingStatus::AW_EMERGENCY:
+        publishCommand(static_cast<std::underlying_type<CommandState>::type>(CommandState::ERROR));
+        RCLCPP_ERROR(this->get_logger(), "AW emergency");
+        break;
+      // AWが停留所外などではSEND_ZEROを発出し、設備連携結果はFAILで返す
+      case InParkingStatus::AW_OUT_OF_PARKING:
+      case InParkingStatus::AW_UNAVAILABLE:
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp:136]:
```suggestion
      case InParkingStatus::AW_UNAVAILABLE:
      case InParkingStatus::NONE:
```
https://github.com/eve-autonomy/cargo_loading_service/pull/2#discussion_r1114165828
のスレッドでのNONEの移動先はこちらで良いと思います。
この場合、設備に終了通知0b00を送り即時終了します。

**Review by yn-mrse — CHANGES_REQUESTED** [review]

File: src/cargo_loading_service.cpp [hunk 19]
```diff
@@ -12,150 +12,178 @@
...
-      if (state.id.compare(facility_info_.id) == 0) {
-        facility_info_.approval = state.approval;
-      }
-    }
+  aw_state_ = msg->aw_state;
+
+  RCLCPP_DEBUG_THROTTLE(
+    this->get_logger(), *this->get_clock(), 1000 /* ms */, "inParkingStatus: %s",
+    rosidl_generator_traits::to_yaml(*msg).c_str());
+}
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp:173]: aw_state_の値が一定期間取れないときにaw_state = AW_EMERGENCYの処理（設備にエラー通知）が必要かと思います。
この関数のtimeout監視をして、設備連携中（!infra_approval_のとき）にtimeoutが発生した場合にはErrorに入れるなどのFailsafeが期待されるはずです。

ただし今回は時間もないため制限事項として管理する方針でご提案ください。

File: src/cargo_loading_service.cpp [hunk 20]
```diff
@@ -12,150 +12,178 @@
...
+    service_result_ = ExecuteInParkingTask::Response::SUCCESS;
+    timer_->cancel();
   }
-  RCLCPP_DEBUG(this->get_logger(), "Subscribed /in_parking/state:%s",
-               rosidl_generator_traits::to_yaml(*msg).c_str());
 }
 
-void CargoLoadingService::onCargoLoadingState(
-  const v2i_interface_msgs::msg::InfrastructureStateArray::ConstSharedPtr msg)
+void CargoLoadingService::onInParkingStatus(const InParkingStatus::ConstSharedPtr msg)
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp:168]: aw_state_のcallbackが一定期間呼ばれないときにaw_state_ = AW_EMERGENCYのときの処理（設備にエラー通知）が必要かと思います。
この関数のtimeout監視をして、設備連携中（!infra_approval_のとき）にtimeoutが発生した場合にはErrorに入れるなどのFailsafeが期待されるはずです。

ただし今回は時間もないため制限事項として管理する方針でご提案ください。

File: include/cargo_loading_service/cargo_loading_service.hpp [hunk 21]
```diff
@@ -35,43 +37,50 @@ class CargoLoadingService : public rclcpp::Node
...
-  std::mutex mutex_cargo_loading_state_;
-  std::mutex mutex_parking_state_;
-  struct FacilityInfo {
-    std::string id;
-    bool approval;
-  } facility_info_;
-  int32_t aw_state_;
-  std::chrono::nanoseconds command_pub_sleep_time_;
-  double cargo_loading_command_pub_hz_;
+  static constexpr char COMMAND_TYPE[] = "eva_beacon_system";
```
Comment (yn-mrse) [inline include/cargo_loading_service/cargo_loading_service.hpp]:
```suggestion
```

File: src/cargo_loading_service.cpp [hunk 22]
```diff
@@ -12,150 +12,178 @@
...
 
-  // Initialize class variables
-  facility_info_.id = "";
-  {
-    std::lock_guard<std::mutex> lock(mutex_cargo_loading_state_);
-    facility_info_.approval = false;
-  }
+  InfrastructureCommand command;
+  command.stamp = stamp;
+  command.type = COMMAND_TYPE;
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
```

File: include/cargo_loading_service/cargo_loading_service.hpp [hunk 23]
```diff
@@ -35,43 +37,50 @@ class CargoLoadingService : public rclcpp::Node
...
 
-  // Callback group
-  rclcpp::CallbackGroup::SharedPtr callback_group_service_;
-  rclcpp::CallbackGroup::SharedPtr callback_group_subscription_;
+  // variable
+  std::string facility_id_;
```
Comment (yn-mrse) [inline include/cargo_loading_service/cargo_loading_service.hpp]:
```suggestion
  uint8_t infra_id_;
```
facility vs infraで使い分けに意味があまりないと思いますので、主流なinfraのほうに修正します。

File: src/cargo_loading_service.cpp [hunk 24]
```diff
@@ -12,150 +12,178 @@
...
+  // サービスのresult更新
+  response->state = service_result_;
 
-  {
-    std::lock_guard<std::mutex> lock(mutex_cargo_loading_state_);
-    facility_info_.approval = false;
-  }
+  // 各変数の初期化
+  service_result_ = ExecuteInParkingTask::Response::NONE;
+  facility_id_ = "";
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
  infra_id_ = InfrastructureState::INVALID_ID;
```

File: src/cargo_loading_service.cpp [hunk 25]
```diff
@@ -12,150 +12,178 @@
...
-  // Initialize class variables
-  facility_info_.id = "";
-  {
-    std::lock_guard<std::mutex> lock(mutex_cargo_loading_state_);
-    facility_info_.approval = false;
-  }
+  InfrastructureCommand command;
+  command.stamp = stamp;
+  command.type = COMMAND_TYPE;
+  command.id = facility_id_;
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
  command.id = infra_id_;
```

File: src/cargo_loading_service.cpp [hunk 26]
```diff
@@ -12,150 +12,178 @@
...
+    rosidl_generator_traits::to_yaml(*msg).c_str());
+}
+
+void CargoLoadingService::onInfrastructureStatus(const InfrastructureStateArray::ConstSharedPtr msg)
+{
+  const auto itr = std::find_if(msg->states.begin(), msg->states.end(), [this](const auto & e) {
+    return e.id == facility_id_;
+  });
+  if (itr != msg->states.end()) {
+    infra_approval_ = msg->states.at(std::distance(msg->states.begin(), itr)).approval;
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
    infra_approval_ = (msg->states.at(std::distance(msg->states.begin(), itr)).status == CommandState::ERROR);
```

File: src/cargo_loading_service.cpp [hunk 27]
```diff
@@ -12,150 +12,178 @@
...
+  RCLCPP_DEBUG_THROTTLE(
+    this->get_logger(), *this->get_clock(), 1000 /* ms */, "inParkingStatus: %s",
+    rosidl_generator_traits::to_yaml(*msg).c_str());
+}
+
+void CargoLoadingService::onInfrastructureStatus(const InfrastructureStateArray::ConstSharedPtr msg)
+{
+  const auto itr = std::find_if(msg->states.begin(), msg->states.end(), [this](const auto & e) {
+    return e.id == facility_id_;
+  });
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
  if (infra_id_ == InfrastructureState::INVALID_ID) {
    return;
  }
  const auto itr = std::find_if(msg->states.begin(), msg->states.end(), [this](const auto & e) {
    return e.id == infra_id_;
  });
```

**Review by yn-mrse — CHANGES_REQUESTED** [review]
https://github.com/eve-autonomy/v2i_interface_msgs/pull/2 を反映した場合の追加修正提案です。

File: src/cargo_loading_service.cpp (hunk 16 above)
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
```
https://github.com/eve-autonomy/cargo_loading_service/pull/2#discussion_r1114177150 にあわせて削除のコメントをしていましたが誤って消えたため再度コメントします。

**Review by yn-mrse — CHANGES_REQUESTED** [review]

File: src/cargo_loading_service.cpp (hunk 12 above)
Comment (h-ohta) [reply src/cargo_loading_service.cpp]: たしかにそうですね。修正します

File: src/cargo_loading_service.cpp (hunk 13 above)
Comment (h-ohta) [reply src/cargo_loading_service.cpp]: たしかにsleepを入れたほうが良さそう
値を書き換えるわけではなければmutexを入れる必要はないです

File: src/cargo_loading_service.cpp (hunk 27 above)
Comment (h-ohta) [reply src/cargo_loading_service.cpp]:
```suggestion
  const auto itr = std::find_if(msg->states.begin(), msg->states.end(), [this](const auto & e) {
    return e.id == infra_id_;
  });
```
でいいんじゃないでしょうか
INVALID_IDの条件分岐は必要ないと思います

**Review by sfukuta — DISMISSED** [review]

File: src/cargo_loading_service.cpp (hunk 12 above)
Comment (sfukuta) [reply src/cargo_loading_service.cpp]: 反映済みです

File: src/cargo_loading_service.cpp (hunk 13 above)
Comment (sfukuta) [reply src/cargo_loading_service.cpp]: sleepは、対応済み。mutexは不要となります。

File: src/cargo_loading_service.cpp (hunk 20 above)
Comment (yn-mrse) [reply src/cargo_loading_service.cpp:168]: [T4PB-26456(TIER IV INTERNAL LINK)](https://tier4.atlassian.net/browse/T4PB-26456)にて管理とし、このスレッドはCloseします。

File: src/cargo_loading_service.cpp [hunk 28]
```diff
@@ -12,150 +12,180 @@
...
-  const in_parking_msgs::srv::ExecuteInParkingTask::Response::SharedPtr response)
+  const ExecuteInParkingTask::Request::SharedPtr request,
+  const ExecuteInParkingTask::Response::SharedPtr response)
 {
-  using InfrastructureCommand = v2i_interface_msgs::msg::InfrastructureCommand;
-  using InfrastructureCommandArray = v2i_interface_msgs::msg::InfrastructureCommandArray;
-  using InParkingStatus = in_parking_msgs::msg::InParkingStatus;
-  using ExecuteInParkingTaskResponse = in_parking_msgs::srv::ExecuteInParkingTask::Response;
+  // 設備ID取得
+  infra_id_ = request->value;
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
  infra_id_ = request->id;
```

File: src/cargo_loading_service.cpp [hunk 29]
```diff
@@ -12,150 +12,180 @@
...
-    for (const auto & state : msg->states) {
-      if (state.id.compare(facility_info_.id) == 0) {
-        facility_info_.approval = state.approval;
-      }
-    }
+  const auto itr = std::find_if(msg->states.begin(), msg->states.end(), [this](const auto & e) {
+    return e.id == infra_id_;
+  });
+  if (itr != msg->states.end()) {
+    infra_approval_ = (msg->states.at(std::distance(msg->states.begin(), itr)).status == CommandState::ERROR);
```
Comment (yn-mrse) [inline src/cargo_loading_service.cpp]:
```suggestion
    infra_approval_ = (msg->states.at(std::distance(msg->states.begin(), itr)).state == static_cast<uint8_t>(CommandState::ERROR));
```

**Review by yn-mrse — CHANGES_REQUESTED** [review]

**Review by yn-mrse — APPROVED** [review]

**Pull request merged**